connector/connector_ldap.go

```diff
@@ -217,11 +250,108 @@ func (c *LDAPConnector) TrustedEmailProvider() bool {
 	return c.trustedEmailProvider
 }
 
+// A LDAPPool is a Connection Pool for LDAP connections
+// Initialize exported fields and use Get() to get a connection.
+// Use Put() to put it back into the pool.
+type LDAPPool struct {
+	m              sync.Mutex
+	conns          map[*ldap.Conn]struct{}
+	MaxIdleConn    int
+	PoolCheckTimer time.Duration
+	ServerHost     string
+	ServerPort     uint16
+	UseTLS         bool
+	UseSSL         bool
+	TlsConfig      *tls.Config
+}
+
+func (p *LDAPPool) Get(params ...*ldap.Conn) (*ldap.Conn, error) {
+	var conn *ldap.Conn
+
+	p.m.Lock()
+	if len(params) > 0 {
+		conn = params[0]
+	} else {
+		for conn = range p.conns {
+			break
+		}
+	}
+	if conn != nil {
+		_, ok := p.conns[conn]
+		if !ok {
+			p.m.Unlock()
+			return nil, fmt.Errorf("Requested connection does not exist in map.")
+		}
+		delete(p.conns, conn)
+		p.m.Unlock()
+		return conn, nil
+	}
+	p.m.Unlock()
+	return p.ldapConnect()
+}
+
+func (p *LDAPPool) Put(c *ldap.Conn) {
+	p.m.Lock()
+	if p.conns == nil {
+		// First call to Put, initialize map
+		p.conns = make(map[*ldap.Conn]struct{})
+	}
+	if len(p.conns)+1 > p.MaxIdleConn {
+		p.m.Unlock()
+		c.Close()
+		return
+	}
+	p.m.Unlock()
+	// drop to anonymous bind
+	err := c.Bind("", "")
+	if err != nil {
+		// unsupported or disallowed, throw away connection
+		log.Warningf("Unable to re-use LDAP Connection after failure to bind anonymously: %v", err)
+		c.Close()
+		return
+	}
+	p.m.Lock()
+	p.conns[c] = struct{}{}
+	p.m.Unlock()
+}
+
+func (p *LDAPPool) Ping() (int, int) {
+	var conns []*ldap.Conn
+	var alive, killed int
+
+	// Get snapshot of connection-map while holding Lock
+	p.m.Lock()
+	for conn := range p.conns {
+		conns = append(conns, conn)
+	}
+	p.m.Unlock()
+
+	// Iterate over snapshot, Get and ping connections.
+	// Put live connections back into pool, Close dead ones.
+	for _, el := range conns {
+		conn, err := p.Get(el)
```

Comment:
Nice suggestion. My original thought was that having two functions that needed to do the locking right would lead to double the opportunity to at some point do it wrong. But I guess that the complexity of the dual use in the one function does just this by itself. I'll split them up and post a new patch set.

Thanks!